File: android/src/main/java/dev/overair/capacitor/Store.kt
package dev.overair.capacitor

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/** A bundle this device holds on disk. */
data class BundleRecord(
    val id: String,
    val version: String,
    val path: String,
    val checksum: String,
    val size: Long,
    /** The native build it was unpacked for. */
    val nativeBuild: String,
) {
    fun toJson(): JSONObject = JSONObject()
        .put("id", id).put("version", version).put("path", path)
        .put("checksum", checksum).put("size", size).put("nativeBuild", nativeBuild)

    companion object {
        fun from(json: JSONObject): BundleRecord = BundleRecord(
            id = json.getString("id"),
            version = json.optString("version"),
            path = json.getString("path"),
            checksum = json.optString("checksum"),
            size = json.optLong("size"),
            nativeBuild = json.optString("nativeBuild"),
        )
    }
}

/**
 * Plugin state, in native SharedPreferences.
 *
 * Native storage rather than anything the web layer can reach, and that is
 * the point: a bundle must not be able to edit the facts that decide whether
 * it is allowed to run.
 */
class Store(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("OverairState", Context.MODE_PRIVATE)

    /** Generated once and kept forever. Deliberately not a device identifier:
     *  a reinstall is a new install, which is the correct behaviour. */
    val installId: String
        get() {
            prefs.getString(INSTALL_ID, null)?.let { return it }
            val fresh = UUID.randomUUID().toString()
            prefs.edit().putString(INSTALL_ID, fresh).apply()
            return fresh
        }

    var storedBuild: String?
        get() = prefs.getString(NATIVE_BUILD, null)
        set(value) = prefs.edit().putString(NATIVE_BUILD, value).apply()

    var active: BundleRecord?
        get() = record(ACTIVE)
        set(value) = put(ACTIVE, value)

    var next: BundleRecord?
        get() = record(NEXT)
        set(value) = put(NEXT, value)

    /** Set when a bundle is handed to the webview, cleared by notifyReady().
     *  A launch that finds it still set knows the last one never came back. */
    var pending: Boolean
        get() = prefs.getBoolean(PENDING, false)
        set(value) = prefs.edit().putBoolean(PENDING, value).apply()

    /** Reported to the caller once, so the SDK can send a FAILED event for a
     *  rollback that happened before any JavaScript existed to notice it. */
    var rolledBackId: String?
        get() = prefs.getString(ROLLED_BACK, null)
        set(value) = prefs.edit().putString(ROLLED_BACK, value).apply()

    fun quarantined(): List<String> {
        val raw = prefs.getString(BAD, null) ?: return emptyList()
        return runCatching {
            val array = JSONArray(raw)
            (0 until array.length()).map { array.getString(it) }
        }.getOrDefault(emptyList())
    }

    fun quarantine(id: String) {
        val current = quarantined()
        if (current.contains(id)) return
        // Bounded: this list travels in every check request, and a device
        // that keeps failing must not grow its own payload without limit.
        val next = (current + id).takeLast(20)
        prefs.edit().putString(BAD, JSONArray(next).toString()).apply()
    }

    /** Forget every bundle. Used when a store update lands. */
    fun forgetBundles() {
        prefs.edit().remove(ACTIVE).remove(NEXT).putBoolean(PENDING, false).apply()
    }

    private fun record(key: String): BundleRecord? {
        val raw = prefs.getString(key, null) ?: return null
        // A corrupt record is the same as no record: the bundle it names is
        // unverifiable, so the safe reading is that we have nothing.
        return runCatching { BundleRecord.from(JSONObject(raw)) }.getOrNull()
    }

    private fun put(key: String, value: BundleRecord?) {
        val editor = prefs.edit()
        if (value == null) editor.remove(key) else editor.putString(key, value.toJson().toString())
        editor.apply()
    }

    private companion object {
        const val INSTALL_ID = "install_id"
        const val NATIVE_BUILD = "native_build"
        const val ACTIVE = "active"
        const val NEXT = "next"
        const val PENDING = "pending"
        const val BAD = "quarantined"
        const val ROLLED_BACK = "rolled_back"
    }
}

File: ios/Sources/OverairPlugin/Store.swift
import Foundation

/// Plugin state, in UserDefaults under its own suite.
///
/// Native storage rather than anything the web layer can reach, and that is
/// the point: a bundle must not be able to edit the facts that decide whether
/// it is allowed to run.
public final class Store {

    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private enum Key {
        static let installId = "overair.install_id"
        static let nativeBuild = "overair.native_build"
        static let active = "overair.active"
        static let next = "overair.next"
        static let previous = "overair.previous"
        static let pending = "overair.pending"
        static let bad = "overair.quarantined"
        static let rolledBack = "overair.rolled_back"
    }

    /// Generated once and kept forever. Deliberately not a device identifier:
    /// a reinstall is a new install, which is the correct behaviour.
    public var installId: String {
        if let existing = defaults.string(forKey: Key.installId) { return existing }
        let fresh = UUID().uuidString.lowercased()
        defaults.set(fresh, forKey: Key.installId)
        return fresh
    }

    public var storedBuild: String? {
        get { defaults.string(forKey: Key.nativeBuild) }
        set { defaults.set(newValue, forKey: Key.nativeBuild) }
    }

    public var active: BundleRecord? {
        get { record(Key.active) }
        set { put(newValue, at: Key.active) }
    }

    public var next: BundleRecord? {
        get { record(Key.next) }
        set { put(newValue, at: Key.next) }
    }

    /// What was active before the current one. A bundle that breaks costs the
    /// user one update, not every update they ever took.
    public var previous: BundleRecord? {
        get { record(Key.previous) }
        set { put(newValue, at: Key.previous) }
    }

    /// Set when a bundle is handed to the webview, cleared by `notifyReady()`.
    /// A launch that finds it still set knows the last one never came back.
    public var pending: Bool {
        get { defaults.bool(forKey: Key.pending) }
        set { defaults.set(newValue, forKey: Key.pending) }
    }

    /// Surfaced to the SDK once, so the console learns about a rollback that
    /// happened before any JavaScript existed to notice it.
    public var rolledBackId: String? {
        get { defaults.string(forKey: Key.rolledBack) }
        set { defaults.set(newValue, forKey: Key.rolledBack) }
    }

    public var quarantined: [String] {
        defaults.stringArray(forKey: Key.bad) ?? []
    }

    public func quarantine(_ id: String) {
        var current = quarantined
        guard !current.contains(id) else { return }
        current.append(id)
        // Bounded: this list travels in every check request, and a device
        // that keeps failing must not grow its own payload without limit.
        defaults.set(Array(current.suffix(20)), forKey: Key.bad)
    }

    /// Forget every bundle. Used when a store update lands.
    public func forgetBundles() {
        defaults.removeObject(forKey: Key.active)
        defaults.removeObject(forKey: Key.next)
        defaults.removeObject(forKey: Key.previous)
        defaults.set(false, forKey: Key.pending)
    }

    private func record(_ key: String) -> BundleRecord? {
        guard let data = defaults.data(forKey: key) else { return nil }
        // A corrupt record is the same as no record: the bundle it names is
        // unverifiable, so the safe reading is that we have nothing.
        return try? JSONDecoder().decode(BundleRecord.self, from: data)
    }

    private func put(_ value: BundleRecord?, at key: String) {
        guard let value, let data = try? JSONEncoder().encode(value) else {
            defaults.removeObject(forKey: key)
            return
        }
        defaults.set(data, forKey: key)
    }
}
